### PR TITLE
[WIP][MM-18955] use typescript in components/common

### PR DESCRIPTION
**Pull request by Mrigank11**
#### Summary

Migrate 'components/common' module and associated tests to TypeScript
#### Ticket Link

Fixes mattermost/mattermost-server#12472

### CODE REVIEW

**Comment by Mrigank11** [comment]
The version of `react-bootstrap` used ([here](https://github.com/mattermost/mattermost-webapp/blob/master/package.json#L52)) does not contain typings and the ones at `@types/react-bootstrap` are not updated anymore (like OverlayTrigger definition does not include `className` which is required at [comment_icon.tsx#L74](https://github.com/mattermost/mattermost-webapp/blob/d9d8a614150ce51f43ade7662a1d939a2bc36542/components/common/comment_icon.tsx#L74))

Should I update `react-bootstrap` in `package.json`?

**Comment by reflog** [comment]
@Mrigank11 - we don't want to update react-bootstrap at this point. For missing stuff - please create a type definition.

**Comment by Mrigank11** [comment]
@reflog I have added the `d.ts` files but there's one more issue: https://github.com/benmosher/eslint-plugin-import/issues/1341

Permission to use https://github.com/alexgorbatchev/eslint-import-resolver-typescript?

**Comment by reflog** [comment]
> @reflog I have added the `d.ts` files but there's one more issue: [benmosher/eslint-plugin-import#1341](https://github.com/benmosher/eslint-plugin-import/issues/1341)
> 
> Permission to use [alexgorbatchev/eslint-import-resolver-typescript](https://github.com/alexgorbatchev/eslint-import-resolver-typescript)?

@jespino - what do you think?

File: components/common/back_button.tsx [hunk 1]
```diff
@@ -6,7 +6,12 @@ import React from 'react';
...
 import {Link} from 'react-router-dom';
 
-export default class BackButton extends React.PureComponent {
+type Props = {
+    url: string;
+    onClick?: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void;
+}
+
+export default class BackButton extends React.PureComponent<Props> {
     static propTypes = {
```
Comment (jespino) [inline components/common/back_button.tsx]: you no longer need the propTypes

File: components/common/back_button.tsx [hunk 2]
```diff
@@ -38,7 +43,7 @@ export default class BackButton extends React.PureComponent {
                         id='generic_icons.back'
                         defaultMessage='Back Icon'
                     >
-                        {(title) => (
+                        {(title: any) => (
```
Comment (jespino) [inline components/common/back_button.tsx]: Title here should be always a string

File: components/common/comment_icon.tsx [hunk 3]
```diff
@@ -11,7 +11,16 @@ import {localizeMessage} from 'utils/utils.jsx';
...
+    extraClass: string;
+    postId: string;
+    location: string;
+    searchStyle: string;
+    commentCount: number;
+    handleCommentClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
+}
+
+export default class CommentIcon extends React.PureComponent<Props> {
     static propTypes = {
```
Comment (jespino) [inline components/common/comment_icon.tsx]: You no longer need propTypes here

File: components/common/comment_icon.tsx [hunk 4]
```diff
@@ -11,7 +11,16 @@ import {localizeMessage} from 'utils/utils.jsx';
 
 import ReplyIcon from 'components/widgets/icons/reply_icon';
 
-export default class CommentIcon extends React.PureComponent {
+type Props = {
+    extraClass: string;
+    postId: string;
+    location: string;
```
Comment (jespino) [inline components/common/comment_icon.tsx]: You are allowing any string as location, and before was only or `Locations.CENTER` or `Locations.SEARCH`.

File: components/common/comment_icon.tsx [hunk 5]
```diff
@@ -11,7 +11,16 @@ import {localizeMessage} from 'utils/utils.jsx';
...
+    searchStyle: string;
+    commentCount: number;
+    handleCommentClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
```
Comment (jespino) [inline components/common/comment_icon.tsx]: This field was required and now is optional, why?

File: components/common/comment_icon.tsx [hunk 6]
```diff
@@ -11,7 +11,16 @@ import {localizeMessage} from 'utils/utils.jsx';
 
 import ReplyIcon from 'components/widgets/icons/reply_icon';
 
-export default class CommentIcon extends React.PureComponent {
+type Props = {
+    extraClass: string;
+    postId: string;
```
Comment (jespino) [inline components/common/comment_icon.tsx]: the `postId` was optional and now is mandatory, why?

File: components/common/site_name_and_description.tsx [hunk 7]
```diff
@@ -5,7 +5,12 @@ import PropTypes from 'prop-types';
...
 import {FormattedMessage} from 'react-intl';
 
-export default class SiteNameAndDescription extends React.PureComponent {
+type Props = {
+    customDescriptionText: string;
+    siteName: string;
+}
+
+export default class SiteNameAndDescription extends React.PureComponent<Props> {
     static propTypes = {
```
Comment (jespino) [inline components/common/site_name_and_description.tsx]: you no longer need propTypes

File: components/common/site_name_and_description.tsx [hunk 8]
```diff
@@ -5,7 +5,12 @@ import PropTypes from 'prop-types';
 import React from 'react';
 import {FormattedMessage} from 'react-intl';
 
-export default class SiteNameAndDescription extends React.PureComponent {
+type Props = {
+    customDescriptionText: string;
```
Comment (jespino) [inline components/common/site_name_and_description.tsx]: `customDescriptionText` was optional and now is mandatory, why?

File: components/post_view/post_info/post_info.jsx [hunk 9]
```diff
@@ -9,10 +9,11 @@ import {OverlayTrigger, Tooltip} from 'react-bootstrap';
 import {Posts} from 'mattermost-redux/constants';
 import * as ReduxPostUtils from 'mattermost-redux/utils/post_utils';
 
+import CommentIcon from 'components/common/comment_icon.jsx';
```
Comment (jespino) [inline components/post_view/post_info/post_info.jsx]: I don't know why you are moving it here, but is not wrong, but you need to remove the extension, because you are renaming the file to `.tsx`

File: components/search_results_item/search_results_item.jsx [hunk 10]
```diff
@@ -8,9 +8,10 @@ import {Posts} from 'mattermost-redux/constants/index';
 import * as ReduxPostUtils from 'mattermost-redux/utils/post_utils';
 import {OverlayTrigger, Tooltip} from 'react-bootstrap';
 
+import CommentIcon from 'components/common/comment_icon.jsx';
```
Comment (jespino) [inline components/search_results_item/search_results_item.jsx]: I don't know why you are moving it here, but is not wrong, but you need to remove the extension, because you are renaming the file to `.tsx`

File: components/should_verify_email/should_verify_email.js [hunk 11]
```diff
@@ -6,6 +6,7 @@ import React from 'react';
 import {FormattedMessage} from 'react-intl';
 
 import BackButton from 'components/common/back_button.jsx';
+
```
Comment (jespino) [inline components/should_verify_email/should_verify_email.js]: unnecesary change

File: types/images.d.ts [hunk 12]
```diff
@@ -2,4 +2,4 @@
 // See LICENSE.txt for license information.
 
 declare module '*.jpg';
-declare module '*.png';
\ No newline at end of file
+declare module '*.png';
```
Comment (jespino) [inline types/images.d.ts]: Unnecessary change

File: types/react-bootstrap/components/OverlayTrigger.d.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (jespino) [inline types/react-bootstrap/components/OverlayTrigger.d.ts]: Why are you adding this files here, instead of using the `@types/react-bootstrap`

Comment (jespino) [reply types/react-bootstrap/components/OverlayTrigger.d.ts]: Nevermind, I saw the comments.

**Comment by jespino** [comment]
> > @reflog I have added the `d.ts` files but there's one more issue: [benmosher/eslint-plugin-import#1341](https://github.com/benmosher/eslint-plugin-import/issues/1341)
> > Permission to use [alexgorbatchev/eslint-import-resolver-typescript](https://github.com/alexgorbatchev/eslint-import-resolver-typescript)?
> 
> @jespino - what do you think?

I'm not 100% sure if that is needed here, I going to try by myself to know exactly the limitations.

File: components/common/back_button.tsx (hunk 2 above)
Comment (Mrigank11) [reply components/common/back_button.tsx]: In that case, I'll have to do something like this:
```tsx
{(...title: Array<string | JSX.Element>) => (
    <span
        id='back_button_icon'
        className='fa fa-1x fa-angle-left'
        title={title[0].toString()}
    />
)}
```
which looks pretty ugly. What do you say?

File: components/common/comment_icon.tsx (hunk 4 above)
Comment (Mrigank11) [reply components/common/comment_icon.tsx]: For that, I'll have to convert `utils/constants.jsx` to typescript because `Locations.CENTER` is a value(read variable) and cannot be used as a type. It needs to be converted to an `enum`.

Should I convert just the location part or leave it for another PR?

**Comment by hanzei** [comment]
Hey @Mrigank11,

Do you have questions about the provided feedback?

**Comment by Mrigank11** [comment]
@hanzei I've already commented my questions above. Waiting for @jespino

File: components/common/comment_icon.tsx (hunk 4 above)
Comment (jespino) [reply components/common/comment_icon.tsx]: I think we can convert the `constants.jsx later`, but for now I would use something like `location: 'CENTER' | 'SEARCH';`

Comment (Mrigank11) [reply components/common/comment_icon.tsx]: I already tried that but it didn't work because `'CENTER'` is not a type.

Comment (jespino) [reply components/common/comment_icon.tsx]: you can define types based on string literals, see for example here: https://github.com/mattermost/mattermost-webapp/blob/master/components/loading_screen.tsx#L8

Comment (Mrigank11) [reply components/common/comment_icon.tsx]: This is what I'm talking about:
```
components/common/comment_icon.tsx:27:9 - error TS2322: Type 'string' is not assignable to type '"CENTER" | undefined'.

27         location: Locations.CENTER,
```
I couldn't find any way to use `"CENTER"` as type and keep `location: Locations.CENTER`

Comment (jespino) [reply components/common/comment_icon.tsx]: Probably you can use directly `"CENTER"` there, that is de default value, and its ok, the constants are great, but in this case the type information is good enough for that.

Comment (jespino) [reply components/common/comment_icon.tsx]: I mean `location: "CENTER"`

File: components/common/back_button.tsx (hunk 2 above)
Comment (jespino) [reply components/common/back_button.tsx]: Can't it be something like:

```tsx
{(title: string | JSX.Element) => (
    <span
        id='back_button_icon'
        className='fa fa-1x fa-angle-left'
        title={title.toString()}
    />
)}
```

Comment (Mrigank11) [reply components/common/back_button.tsx]: Yup, that worked. Thanks!

**Comment by Mrigank11** [comment]
> > > @reflog I have added the `d.ts` files but there's one more issue: [benmosher/eslint-plugin-import#1341](https://github.com/benmosher/eslint-plugin-import/issues/1341)
> > > Permission to use [alexgorbatchev/eslint-import-resolver-typescript](https://github.com/alexgorbatchev/eslint-import-resolver-typescript)?
> > 
> > 
> > @jespino - what do you think?
> 
> I'm not 100% sure if that is needed here, I going to try by myself to know exactly the limitations.

@jespino only this is remaining now. Could you please take a look?

**Comment by devinbinnie** [comment]
@Mrigank11 Can you update this branch to master and fix any merge conflicts? Then I can take a look.

**Comment by hanzei** [comment]
@Mrigank11 Can we help you with fixing the liner issues?

**Comment by M-ZubairAhmed** [comment]
@Mrigank11 if you are caught up, do you mind if i finish this up ? @hanzei

**Comment by jasonblais** [comment]
Hey @Mrigank11, it's been a few weeks, let us know if you're still working on this PR or have any questions?

To follow our [inactive contribution process](https://developers.mattermost.com/contribute/getting-started/inactive-contributions/), this PR is eligible to be assumed by another community member in a couple of days.

@M-ZubairAhmed let's wait a couple of days to hear from @Mrigank11 - if we haven't heard back, you can take it over.

**Comment by Mrigank11** [comment]
@M-ZubairAhmed you can take it over.

**Comment by hanzei** [comment]
Closing this PR. @M-ZubairAhmed will continue to work on https://github.com/mattermost/mattermost-server/issues/12472
